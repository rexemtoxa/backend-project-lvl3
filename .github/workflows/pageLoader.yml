# This workflow will do a clean install of node dependencies, build the source code and run tests across different versions of node
# For more information see: https://help.github.com/actions/language-and-framework-guides/using-nodejs-with-github-actions

name: pageLoader

on:
  push:
    branches: [ master ]

jobs:
  test:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        node-version: [13.x]
    steps:
    - uses: actions/checkout@v2
    - name: Use Node.js ${{ matrix.node-version }}
      uses: actions/setup-node@v1
      with:
        node-version: ${{ matrix.node-version }}
    - run: npm ci
    - run: npm test
    - run: npm run build
    - run: npm run test-publish
      env:
        CI: true

  coverage:
    needs: [ test ]
    strategy:
      matrix:
        node-version: [13.x]
    name: coverage
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@master
    - uses: actions/setup-node@master
      with:
        node-version: ${{ matrix.node-version }}
    - run: npm ci
    - uses: paambaati/codeclimate-action@v2.5.5
      env:
        CC_TEST_REPORTER_ID: ${{secrets.codeClimate}}
      with:
        coverageCommand: npm run test --coverage --watchAll=false
        coverageLocations:
          "${{github.workspace}}/test-coverage/lcov.info:lcov"
  

  publish:
    needs: [ test ]
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@master
      - uses: actions/setup-node@master
        with:
          node-version: 13
          registry-url: https://registry.npmjs.org/
      - name: build package
        - run: npm ci
        - run: npm run build
      - name: Register Token
        run: |
          echo "//registry.npmjs.org/:_authToken=$NODE_AUTH_TOKEN" > /home/runner/work/_temp/.npmrc
          echo "_auth=$NODE_AUTH_TOKEN" >>  /home/runner/work/_temp/.npmrc
          echo "email=<rehemtoha@gmail.com>" >>  /home/runner/work/_temp/.npmrc
          echo "always-auth=true" >>  /home/runner/work/_temp/.npmrc
        env:
          NODE_AUTH_TOKEN: ${{secrets.npm}}
      - name: Publish stable
        run: npm publish
