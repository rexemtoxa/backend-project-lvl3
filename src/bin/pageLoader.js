#!/usr/bin/env node
import program from 'commander';
import { version } from '../../package.json';
import pageLoader from '..';

program
  .version(version)
  .description('load and save page to the output directory')
  .option('-o, --output [output folder] ', 'Output folder', process.cwd())
  .arguments('<url>')
  .action((url) => (
    pageLoader(url, program.output)))
  .parse(process.argv);
