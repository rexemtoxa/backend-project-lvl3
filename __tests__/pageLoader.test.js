import nock from 'nock';
import { promises as fs } from 'fs';
import os from 'os';
import path from 'path';
import debug from 'debug';
import pageLoader from '../src';

const logger = debug('nock');

const genNameTempDir = () => path.join(os.tmpdir(), 'page-loader-');
const getPathToFixture = (pathToFile) => path.join(__dirname, '__fixtures__', ...pathToFile);
let tempDir;

describe('get http request to page and save html', () => {
  // Arrange
  beforeEach(async () => {
    tempDir = await fs.mkdtemp(genNameTempDir());
  });

  beforeAll(async () => {
    nock.disableNetConnect();
    nock('http://test.positive.com')
      .get('/')
      .replyWithFile(200, getPathToFixture(['pageWithOutLocalResources.html']))
      .get('/unknown/page')
      .reply(404)
      .log(logger);
  });
  test('save page if response returns status code 200', async () => {
    // Act
    await pageLoader('http://test.positive.com', tempDir);
    // Assert
    const savedPage = await fs.readFile(path.join(tempDir, 'test-positive-com.html'), 'utf-8');
    expect(savedPage).toMatchSnapshot();
  });

  test('page should not be saved, if request fails', async () => {
    // Act
    await expect(pageLoader('http://test.positive.com/unknown/page', tempDir)).rejects.toThrow('Request failed with status code 404');
    // Assert
    await expect(fs.access(path.join(tempDir, 'test-positive-com-unknown-page.html'))).rejects
      .toThrow(/ENOENT: no such file or directory, access '\/tmp\/page-loader-.*\/test-positive-com-unknown-page.html'$/);
  });
});

describe('load page with local resources and change the src', () => {
  // Arrange
  beforeEach(async () => {
    tempDir = await fs.mkdtemp(genNameTempDir());
  });

  beforeAll(async () => {
    nock.disableNetConnect();
    nock('http://check.amount.local.files.com')
      .get('/with/local/files')
      .replyWithFile(200, getPathToFixture(['pageWithLocalResources', 'pageWithLocalResources.html']))
      .get('/assets/logo.jpg')
      .replyWithFile(200, getPathToFixture(['pageWithLocalResources', 'assets', 'logo.jpg']))
      .get('/assets/index.js')
      .replyWithFile(200, getPathToFixture(['pageWithLocalResources', 'assets', 'index.js']))
      .get('/assets/style.css')
      .replyWithFile(200, getPathToFixture(['pageWithLocalResources', 'assets', 'style.css']))
      .get('/assets/error')
      .replyWithError({ message: 'went smth wrong', status: 500 })
      .log(logger);
  });
  const getOutputPath = (fileName) => path.join(
    tempDir, 'check-amount-local-files-com-with-local-files_files',
    fileName,
  );
  test('check loading page with local files', async () => {
    // Act
    await pageLoader('http://check.amount.local.files.com/with/local/files', tempDir);
    const expectImage = await fs.readFile(getPathToFixture(['pageWithLocalResources', 'assets', 'logo.jpg']));
    const actualScript = await fs.readFile(getOutputPath('index.js'), 'utf-8');
    const actualStyle = await fs.readFile(getOutputPath('style.css'), 'utf-8');
    const actualImage = await fs.readFile(getOutputPath('logo.jpg'));
    // Assert
    expect(actualScript).toMatchSnapshot();
    expect(actualStyle).toMatchSnapshot();
    expect(actualImage).toEqual(expectImage);
  });
});

describe('error scenarios', () => {
  test('not have a permission te create a dir', async () => {
    nock('http://test.positive.com')
      .get('/').reply(200);
    await expect(pageLoader('http://test.positive.com/', '/path_new_dir')).rejects.toThrow("EACCES: permission denied, mkdir '/path_new_dir'");
  });
});
