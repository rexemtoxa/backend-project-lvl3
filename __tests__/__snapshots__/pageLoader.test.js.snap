// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get http request to page and save html save page if response returns status code 200 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
   <head>
      <meta charset=\\"UTF-8\\">
      <meta name=\\"viewport\\" content=\\"width=720, initial-scale=1.0\\">
      <title>Document</title>
   </head>
   <body>
      <div>
         <h1>Выстрелить себе в ногу</h1>
      </div>
      <div>
         <ul>
            <li> <b>Параллельный Euclid</b>: Вы простреливаете себе чью-то чужую ногу.
            </li>
            <li> <b>Язык управления заданиями IBM/370</b>: Вы посылаете свою ногу в управляющую информационную систему с приложением документа на 400 страниц, точно описывающего, как именно вы хотите быть простреленным. Через три года ваша нога возвращается с глубоким ожогом.
            </li>
            <li> <b>Paradox</b>: Не только вы можете прострелить себе ногу, но и ваши пользователи.
            </li>
            <li> <b>Revelation</b>: Вы уверены, что сможете прострелить себе ногу, как только разберетесь, для чего служат эти маленькие хорошенькие пульки.
            </li>
            <li> <b>Ассемблер</b>: Вы пытаетесь прострелить себе ногу, но обнаруживаете, что прежде вам придется изобрести пистолет, пулю, курок и вашу ногу. После чего вы нажимаете на курок, пистолет издаёт пару писков, после чего выдаёт.
            </li>
            <li> <b>Ассемблер v2</b>: Вы пытаетесь прострелить себе ногу, но наступаете ею на грабли.
            </li>
            <li> <b>HQ9+</b>: У вас нет ни пистолета, ни «ноги».
            </li>
            <li> <b>VHDL</b>: Вы аккуратно описываете компоненты «нога», «рука», «пистолет» и «пуля». Любовно их отлаживаете и моделируете по отдельности. После синтеза всей системы обнаруживается, что нога и пистолет активируются на чётных тактах сигнала синхронизации, а рука и пуля на нечётных.
            </li>
            <li> <b>Whitespace</b>: &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
            </li>
            <li> <b>Redcode</b>: Вы стреляете себе в ногу, но у пули на сей счет иные планы. Она постоянно ищет себя, стреляет себе в ногу, удивляется результату…
            </li>
            <li> <b>ABAP/4</b>: В стандартном пистолете предусмотрена возможность стрелять из-под колена в белку на расстоянии 3&nbsp;км, но выстрелить в ногу нельзя. После годовой доработки можно стрелять в ногу. На второй минуте использования выясняется, что стрелять нужно было, только если нога левая, а четвертая.
            </li>
            <li> <b>ColdFusion</b>: Программа снимает за вас ваши туфли, заряжает ваш пистолет и назначает вам прием к доктору. Вы успешно стреляете себе в ногу, но вам никто не верит.
            </li>
            <li> <b>XSLT</b>: Вы гуглите весь день, пытаясь узнать, как же выстрелить себе в ногу. Узнав решение, вы не понимаете его синтаксис и берете чужой шаблон. Он не работает.
            </li>
            <li> <b>Lua</b>: Если разработчик пистолета предусмотрел возможность выстрела в ногу — достаточно вызвать одну функцию. Если нет&nbsp;— хоть убейтесь, но ничего не выйдет.
            </li>
            <li> <b>Squirrel</b>: Намаявшись с Lua, вы прогрызаете ногу той самой белкой, глаз которой вы отстрелили из пистолета «ABAP».
            </li>
            <li> <b>Apple Script</b>: Вы простреливаете себе ногу, но перед этим вам придётся сделать автоматическую запись ваших действий в компилятор&nbsp;— прострел собственной ноги, и в итоге вы автоматически простреливаете себе ногу вторично.
            </li>
            <li> <b>FoxPro</b>: Вы хотите выстрелить себе в ногу, но обнаруживаете, что нога уже прострелена, и вам приходится искать пользователя, который это сделал.
            </li>
            <li> <b>MUMPS/M</b>: Вы используете переменную glo для ссылки на глобаль с состоянием своего тела, потом обращаетесь к функции, где производится выстрел в ногу и сохраняете результат в индекс ноги. Но в метке кто-то уже использует переменную с тем же именем для получения данных о пистолете, не ньюкнув ее предварительно. В результате вы в полном порядке, а у пистолета прострелена правая нога.
            </li>
            <li> <b>UART</b>: У вашего пистолета такой короткий ствол, что патрон просто торчит из дула. Вам приходится расковыривать ногу, причём так, чтобы именно выстрел пластмассовой пульки нанёс ноге рану.
            </li>
            <li> <b>Go</b>: Вы создаете пулю и запускаете ее в полет. Тем временем создаете ногу, пистолет, и сообщаете пуле, что она должна лететь от пистолета к ноге. При подлете к ноге пуля исчезает, так как ногу простреливать запрещено.
            </li>
            <li> <b>GML (Game Maker Language)</b>: Вы покупаете Desert Eagle Custom за 2999,99$, пишете instance_create(x,y,obj_bullet) и задаёте на различных вопрос: «пачиму пуля нилитит кнаге7».
            </li>
            <li> <b>DOS Batch</b>: Вы прицеливаетесь в ногу из пистолета и нажимаете на курок, но вашу ногу лишь обдувает теплым ветерком.
            </li>
            <li> <b>Машинный язык</b>: Перед тем, как выстрелить в ногу, вам нужно расщепить пистолет на атомы, а затем снова его собрать. После этого вы обнаруживаете, что для пуль требуется специальный оператор, поэтому пистолет не работает.
            </li>
            <li> <b>PL/I</b>: У вас есть 6,752 различных пистолетов. Пока вы выбираете один, чтобы выстрелить себе в ногу, вы умираете от голода.
            </li>
            <li> <b>Qt</b>: Вы пытаетесь прострелить себе ногу но понимаете что у вас нет нужного сигнала, вы наследуете себя от класса \\"QЧиловек\\", приписываете метод, делаете emit сигнала, ставите нужный флаг который даёт вам возможность стрелять... но пистолет не стреляет. Вы наследуете \\"QПистолет\\", прописываете слот, ставите флаг IsShootable, и обнаруживаете что нога не может быть простреленной... вы в ужасе идёте наследовать и её, устанавливаете ещё 10 флагов. Теперь всё готово к прострелу, вы стреляете и в результате получаете Runtime error
            </li>
            </li><li> <b>SQL</b>: INSERT INTO нога (ступня) SELECT патроны AS пуля FROM пистолет WHERE курок = 'СПУЩЕН'; COMMIT;
            </li><li> <b>SQL v2</b>: SELECT @ammo:=bullet FROM gun WHERE trigger = ‘PULLED’; INSERT INTO leg (foot) VALUES (@ammo);
            </li><li> <b>CSS</b>: Переложив пистолет в левую руку, вы понимаете, что держите его не за рукоять, а за дуло. Вы переворачиваете его и понимаете, что теперь он смотрит стволом в сторону вашей головы. Ещё немного изменив стили, вы понимаете, что всё теперь как надо, но только он находится на 15 пикселей левее вашей руки.
            </li><li> <b>node.JS</b>: Вы начинаете асинхронно стрелять из асинхронных рук в асинхронные ноги, асинхронно не попадаете и запутываетесь в этой каше.
            </li><li> <b>Objective-C/Cocoa</b>: Вы строите стильный интерфейс, соответствующий всем канонам Apple. Вы два раза копируете и ногу, и пистолет, чтобы они работали под архитектурами как Intel, так и PowerPC. Вы удостоверяетесь, что пистолет можно зарядить с помощью всего одного клика мышкой. Когда всё уже готово, вы понимаете, что при нажатии на спуск пистолет не стреляет. Перебрав весь код вашей программы, вы обнаруживаете, что забыли указать, что графический объект спускового крючка&nbsp;— и есть спусковой крючок.
            </li><li> <b>ASP.NET</b>: Вы находите пистолет, но он разваливается. Вы его собираете, но он снова разваливается. Вы пытаетесь использовать платформу .GUN, но она разваливается. Вы режете себе ногу
            </li><li> <b>FORTRAN</b>: Итеративно вы простреливаете каждый палец ноги, а когда пальцы заканчиваются, загружаете следующую ногу и продолжаете. Когда пули заканчиваются, а пальцы давно  потеряли всякое сходство с пальцами, вы все равно продолжаете попытки прострелить себе ноги, потому что не имеете возможности обрабатывать исключения.
            </li><li> <b>Modula2</b>: Поняв, что с помощью этого языка ничего толком сделать нельзя, вы простреливаете себе голову.
            </li><li> <b>Rust</b>: Вы покупаете пистолет, который может стрелять только при ночной сборке. Вы пытаетесь выстрелить в ногу, однако выясняется, что нога заимствована и неизменяема, а пуля не может быть перемещена. В конечном счёте вы игнорируете все проверки, после чего пистолет взрывается, а ноги остаются атрофированными по сравнению с торсом из-за <a rel=\\"nofollow\\" class=\\"external text\\" href=\\"https://twitter.com/sangwhanmoon/status/664307517581733888\\">пропуска дня ног</a> в спортзале.
            </li><li> <b>Ada</b>: После тщательной сборки пакета «нога», вы пытаетесь одновременно зарядить пистолет, нажать курок, завопить и выстрелить себе в ногу. В ходе попыток обнаруживается, однако, что вас постигла неудача, так как ваша нога оказалась неверного типа.
            </li><li> <b>Кобол</b>: Наводите ПИСТОЛЕТ COLT 45 на НОГА.СТОПА, ТОГДА помещаете РУКА.КИСТЬ.ПАЛЕЦ на ПИСТОЛЕТ.КРЮЧОК и ПОЖАЛУЙСТА, НАЖИМАЕТЕ. ТОГДА возвращаете ПИСТОЛЕТ в КОБУРУ. ПОЖАЛУЙСТА, ПРОВЕРЬТЕ, нуждаются ли шнурки в повторном связывании.
            </li><li> <b>Basic</b>: Вы простреливаете себе ногу из водяного пистолета. В расширенных реализациях языка продолжайте, пока вся нижняя часть тела не промокнет. В реализациях с возможностью делать вставки из ассемблера вы используете ранее созданные типы и уже в ассемблерной части простреливаете ногу дробовиком.
            </li><li> <b>Visual Basic</b>: Вы прострелите себе ногу. Однако в процессе реализации вы получите столько удовольствия, что подобные мелочи вас не озаботят.
            </li><li> <b>Visual BASIC.NET</b>: Вы тратите целое состояние на мощнейший, самоочищающийся, самонаводящийся, автоматический пистолет с бесконечным боезапасом и только тогда понимаете, что спуск курка занимает около года (<a rel=\\"nofollow\\" class=\\"external text\\" href=\\"http://www.thealmightyguru.com/Humor/Docs/ShootYourselfInTheFoot.html\\">источник</a>).
            </li><li> <b>HyperTalk</b>: Поместите первую пулю пистолета в ступню левой ноги вас. Сообщите результат.
            </li><li> <b>Smalltalk</b>: Мой нога стреляй себя.
            </li><li> <b>Motif</b>: Вы затрачиваете несколько дней на построение UIL&nbsp;— описания вашей ноги, пули, её траектории и замысловатой резьбы по слоновой кости на рукоятке пистолета. Когда вы наконец-то управились и готовы нажать на курок, затвор заклинивает.
            </li><li> <b>APL</b>: Вы простреливаете себе ногу, затем тратите весь день на то, чтобы представить себе, как это сделать с меньшим количеством символов.
            </li><li> <b>Снобол</b>: Если вам удастся, прострелите себе левую ногу. Если не удастся, то прострелите себе правую ногу.
            </li><li> <b>E2</b>: Вы, используя пистолет, состоящий из двух параллелепипедов кислотного цвета, стреляете в свою невидимую ногу, но пуля вместо ствола пистолета появляется у вас перед глазами, и вы застреваете в ней, поскольку ваша физическая модель представляет из себя цилиндр.
            </li>
         </ul>
      </div>
   </body>
</html>"
`;

exports[`load page with local resources and change the src size of page with assets should equal the size after loading 1`] = `
"// here should be a lot of js code

const fakeFunction = () => '<div><h1>other fake page, and not awesome</h1></div>';
fakeFunction();
"
`;

exports[`load page with local resources and change the src size of page with assets should equal the size after loading 2`] = `
"p {
  border-style: solid;
  border-top: thick double #ff0000;
}
body {
  background-color: black;
  color: aqua;
}"
`;
